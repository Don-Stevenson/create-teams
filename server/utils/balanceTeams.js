function balanceTeams(players, numTeams) {
  const playersThisWeek = [
    ...players.filter(player => player.isPlayingThisWeek),
  ]

  const totalPlayersPlaying = playersThisWeek.length

  const teams = Array.from({ length: numTeams }, () => ({
    players: [],
    totalScore: 0,
    totalGameKnowledgeScore: 0,
    totalGoalScoringScore: 0,
    totalAttackScore: 0,
    totalMidfieldScore: 0,
    totalAttackScore: 0,
    totalDefenseScore: 0,
    fitnessScore: 0,
    genderCount: { male: 0, female: 0, nonBinary: 0 },
  }))

  playersThisWeek.sort((a, b) => {
    if (a.gender !== b.gender) return a.gender.localeCompare(b.gender)
    return (
      b.gameKnowledgeScore +
      b.goalScoringScore +
      b.attackScore +
      b.midfieldScore +
      b.defenseScore +
      b.fitnessScore -
      (b.gameKnowledgeScore +
        b.goalScoringScore +
        b.attackScore +
        b.midfieldScore +
        b.defenseScore +
        b.fitnessScore)
    )
  })

  // Distribute players using a modified serpentine draft method
  for (let i = 0; i < playersThisWeek.length; i++) {
    const player = playersThisWeek[i]
    let teamIndex

    const WEIGHTS = {
      gameKnowledge: 0.2,
      goalScoring: 0.2,
      attack: 0.135,
      midfield: 0.133,
      defense: 0.133,
      fitness: 0.1,
    }

    // add a small amount of noise to help make the teams are not identical each time
    const fudge = score => Math.random() * 0.3 + score - 0.15

    // Calculate player's weighted total score
    player.totalScore =
      fudge(player.gameKnowledgeScore) * WEIGHTS.gameKnowledge +
      fudge(player.goalScoringScore) * WEIGHTS.goalScoring +
      fudge(player.attackScore) * WEIGHTS.attack +
      fudge(player.midfieldScore) * WEIGHTS.midfield +
      fudge(player.defenseScore) * WEIGHTS.defense +
      fudge(player.fitnessScore) * WEIGHTS.fitness

    // Get minimum and maximum team sizes currently
    const minTeamSize = Math.min(...teams.map(t => t.players.length))
    const maxTeamSize = Math.max(...teams.map(t => t.players.length))

    // Find teams that can receive players (lowest size, or lowest size + 1 if necessary)
    let eligibleTeams = teams.filter(
      team => team.players.length === minTeamSize
    )

    // If no teams at minimum size, allow teams at minimum size + 1
    // but only if that wouldn't create more than 1 player difference
    if (eligibleTeams.length === 0 && maxTeamSize - minTeamSize <= 1) {
      eligibleTeams = teams.filter(
        team => team.players.length === minTeamSize + 1
      )
    }

    // Among eligible teams, find the one with lowest total score
    teamIndex = eligibleTeams.reduce(
      (minIndex, team, index, arr) =>
        team.totalScore < arr[minIndex].totalScore ? index : minIndex,
      0
    )

    // Get the index in the original teams array
    teamIndex = teams.findIndex(team => team === eligibleTeams[teamIndex])

    const team = teams[teamIndex]

    team.players.push(player)
    team.totalScore += player.totalScore
    team.totalGameKnowledgeScore += player.gameKnowledgeScore
    team.totalGoalScoringScore += player.goalScoringScore
    team.totalAttackScore += player.attackScore
    team.totalMidfieldScore += player.midfieldScore
    team.totalDefenseScore += player.defenseScore
    team.fitnessScore += player.fitnessScore
    team.genderCount[player.gender]++

    // Validation logging
    // if (i === playersThisWeek.length - 1) {
    //   const teamSizes = teams.map(t => t.players.length)
    //   const finalMaxSize = Math.max(...teamSizes)
    //   const finalMinSize = Math.min(...teamSizes)
    //   console.log(`Final team sizes: ${teamSizes.join(', ')}`)
    //   console.log(`Team size difference: ${finalMaxSize - finalMinSize}`)
    //   console.log(
    //     `Team scores: ${teams.map(t => t.totalScore.toFixed(2)).join(', ')}`
    //   )
    // }
  }

  return { teams, totalPlayersPlaying }
}
export default balanceTeams
