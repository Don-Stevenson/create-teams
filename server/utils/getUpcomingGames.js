// Load environment variables
import { config } from 'dotenv'
import { fileURLToPath } from 'url'
import { dirname, resolve } from 'path'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

// Load .env file from the project root
config({ path: resolve(__dirname, '../../.env') })

// Environment variables
const teamId = process.env.TEAM_ID
const refreshToken = process.env.REFRESH_TOKEN

// Validate environment variables
if (!teamId || !refreshToken) {
  throw new Error(
    'Missing required environment variables: TEAM_ID and REFRESH_TOKEN must be set in .env.local'
  )
}

// GraphQL queries
const ACTIVITIES_QUERY = `
query UpcomingActivitiesList($teamid: String!, $cursor: String, $past: Boolean) {
  activities(input: { teamid: $teamid, cursor: $cursor, past: $past }) {
    edges {
      node {
        _id
        title
        startdate
        meetdate
        cancelled
        isPast
      }
    }
  }
}`

const ACTIVITY_QUERY = `
query GetActivity($teamid: String!, $activityid: String!) {
  activity(input: {teamid: $teamid, activityid: $activityid}) {
    node {
      ...Activity
    }
  }
}

fragment Activity on Activity {
  _id
  title
  startdate
  enddate
  meetdate
  location {
    name
    address
    country
    city
    longitude
    latitude
  }
  timezone
  venue
  inviteType
  invitedGroups
  isRecurring
  cancelled
  isPast
  maxAttendees
  rsvpDueDate
  members {
    _id
    attending
    attendanceDetails {
      category
      note
    }
    seen
    user {
      _id
      fullName
    }
  }
}`

/**
 * Get GraphQL authentication token using refresh token
 * @returns {Promise<string>} The auth token
 */
async function getGraphqlToken() {
  try {
    const response = await fetch('https://auth.heja.io/login/refreshToken', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json;charset=utf-8',
        'heja-api-version': '3',
        'x-app-version-brand': 'web',
        Cookie: `refreshToken=${refreshToken}`,
      },
      body: JSON.stringify({
        credentialType: 'refreshToken',
        refreshToken: 'cookie',
      }),
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return data.data.authToken
  } catch (error) {
    console.error('Error getting GraphQL token:', error)
    throw error
  }
}

/**
 * Make a GraphQL request to Heja API
 * @param {Object} query - The GraphQL query object
 * @returns {Promise<Object>} The response data
 */
async function hejaGraphqlRequest(query) {
  try {
    const token = await getGraphqlToken()

    const response = await fetch('https://web-api.heja.io/graphql', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
        'User-Agent':
          'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:138.0) Gecko/20100101 Firefox/138.0',
      },
      body: JSON.stringify(query),
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return data
  } catch (error) {
    console.error('Error making GraphQL request:', error)
    throw error
  }
}

/**
 * Get upcoming games for a team
 * @param {Object} options - Options object
 * @param {string} options.teamId - The team ID
 * @param {string} [options.cursor] - Pagination cursor
 * @param {boolean} [options.past] - Whether to get past activities
 * @returns {Promise<Array>} Array of upcoming games
 */
async function upcomingGames({ teamId, cursor = null, past = false }) {
  try {
    const response = await hejaGraphqlRequest({
      operationName: 'UpcomingActivitiesList',
      variables: {
        teamid: teamId,
        cursor,
        past,
      },
      query: ACTIVITIES_QUERY,
    })

    return response.data.activities.edges.map(edge => edge.node)
  } catch (error) {
    console.error('Error getting upcoming games:', error)
    throw error
  }
}

/**
 * Get RSVPs for a specific game
 * @param {Object} options - Options object
 * @param {string} options.teamId - The team ID
 * @param {string} options.gameId - The game/activity ID
 * @returns {Promise<Array<string>>} Array of full names of attendees
 */
export async function rsvpsForGame({ teamId, gameId }) {
  try {
    const response = await hejaGraphqlRequest({
      operationName: 'GetActivity',
      variables: {
        activityid: gameId,
        teamid: teamId,
      },
      query: ACTIVITY_QUERY,
    })

    const members = response.data.activity.node.members

    return members
      .filter(member => member.attending)
      .map(member => member.user.fullName)
  } catch (error) {
    console.error('Error getting RSVPs for game:', error)
    throw error
  }
}

// Example usage:
async function getUpcomingGames() {
  try {
    // Get upcoming games
    const games = await upcomingGames({ teamId })
    return games.slice(0, 5)
  } catch (error) {
    console.error('Example error:', error)
  }
}

// console.log('upcomingGames  ', await getUpcomingGames())

export const upcomingGamesList = await getUpcomingGames()
export const rsvpsForGameList = await rsvpsForGame({
  teamId,
  gameId: upcomingGamesList[0]._id,
})

// console.log('rsvpsForGameList', rsvpsForGameList)
