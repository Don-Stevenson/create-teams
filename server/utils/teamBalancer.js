function shuffleArray(array) {
  // Fisher-Yates Shuffle to randomize player order
  for (let i = array.length - 1; i > 0; i--) {
    let j = Math.floor(Math.random() * (i + 1))
    ;[array[i], array[j]] = [array[j], array[i]]
  }
  return array
}

function balanceTeams(players, numTeams) {
  const playersPerTeam = Math.floor(players.length / numTeams)
  const extraPlayers = players.length % numTeams

  const teams = Array.from({ length: numTeams }, () => ({
    players: [],
    totalScore: 0,
    totalAttackScore: 0,
    totalDefenseScore: 0,
    attackerCount: 0,
    defenderCount: 0,
    fitnessScore: 0,
    genderCount: { male: 0, female: 0 },
  }))

  // Shuffle players to ensure variety week-to-week
  const shuffledPlayers = shuffleArray(players)

  // Sort players based on their roles (attackers/defenders)
  const attackers = shuffledPlayers.filter(
    player => player.attackScore > player.defenseScore
  )
  const defenders = shuffledPlayers.filter(
    player => player.defenseScore >= player.attackScore
  )

  // Combine players again for shuffling while maintaining balance
  const combinedPlayers = [...attackers, ...defenders]

  // Split players into male and female for gender balancing
  const males = combinedPlayers.filter(player => player.gender === 'male')
  const females = combinedPlayers.filter(player => player.gender === 'female')

  // Function to find the team that needs a player based on attack/defense balance
  function getBalancedTeam(isAttacker) {
    return teams.reduce(
      (acc, team) => {
        const attackDeficit = team.totalDefenseScore - team.totalAttackScore
        const defenseDeficit = team.totalAttackScore - team.totalDefenseScore
        if (isAttacker && attackDeficit > acc.deficit) {
          return { team, deficit: attackDeficit }
        } else if (!isAttacker && defenseDeficit > acc.deficit) {
          return { team, deficit: defenseDeficit }
        }
        return acc
      },
      { team: teams[0], deficit: Infinity }
    ).team
  }

  // Loop over teams and assign players, balancing gender, roles, and attack/defense scores
  for (let i = 0; i < players.length; i++) {
    // Determine whether to assign an attacker or defender next based on balance
    const isAttackerTurn = i % 2 === 0 ? true : false
    let player = null

    // Pick player based on gender and role
    if (isAttackerTurn) {
      player = attackers.shift()
    } else {
      player = defenders.shift()
    }

    // If no more attackers/defenders are available, continue with remaining players
    if (!player) {
      player = combinedPlayers.shift()
    }

    // Find the best team to balance attack/defense
    const team = getBalancedTeam(isAttackerTurn)

    // Calculate total player score considering attack, defense, and fitness
    const weightedTotalScore =
      player.attackScore * 0.4 +
      player.defenseScore * 0.4 +
      player.fitnessScore * 0.2
    player.totalScore = weightedTotalScore

    // Assign player to the team
    team.players.push(player)
    team.totalScore += player.totalScore
    team.fitnessScore += player.fitnessScore
    team.genderCount[player.gender]++

    // Update attack/defense counts and scores
    if (player.attackScore > player.defenseScore) {
      team.attackerCount++
      team.totalAttackScore += player.attackScore
    } else {
      team.defenderCount++
      team.totalDefenseScore += player.defenseScore
    }
  }

  // Adjust teams for extra players (if the number of players isn't perfectly divisible by teams)
  for (let i = 0; i < extraPlayers; i++) {
    const team = teams[i]
    const player = combinedPlayers.pop() // Add the extra players to the first few teams
    team.players.push(player)
    team.totalScore += player.totalScore
    team.fitnessScore += player.fitnessScore
    team.genderCount[player.gender]++
    if (player.attackScore > player.defenseScore) {
      team.attackerCount++
      team.totalAttackScore += player.attackScore
    } else {
      team.defenderCount++
      team.totalDefenseScore += player.defenseScore
    }
  }

  // Print out the details of each team
  teams.forEach((team, index) => {
    console.log(`\nTeam ${index + 1}:`)
    console.log(
      `Total Score: ${team.totalScore.toFixed(
        2
      )}, Fitness Score: ${team.fitnessScore.toFixed(2)}`
    )
    console.log(
      `Total Attack: ${team.totalAttackScore.toFixed(
        2
      )}, Total Defense: ${team.totalDefenseScore.toFixed(2)}`
    )
    console.log(
      `Males: ${team.genderCount.male}, Females: ${team.genderCount.female}`
    )

    // Sort players by total score before printing
    team.players.sort((a, b) => b.totalScore - a.totalScore)

    team.players.forEach(player => {
      console.log(
        `${player.name} - Attack: ${player.attackScore}, Defense: ${
          player.defenseScore
        }, Fitness: ${player.fitnessScore}, Total: ${player.totalScore.toFixed(
          2
        )}`
      )
    })
  })
  return teams
}

export default balanceTeams
