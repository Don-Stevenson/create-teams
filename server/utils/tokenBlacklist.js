// Token blacklist to track invalidated tokens
const tokenBlacklist = new Set()

export const addToBlacklist = token => {
  tokenBlacklist.add(token)
}

export const isBlacklisted = token => {
  return tokenBlacklist.has(token)
}

export const getBlacklist = () => {
  return tokenBlacklist
}
