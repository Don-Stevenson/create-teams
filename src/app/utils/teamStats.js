const WEIGHTS = {
  gameKnowledge: 0.2,
  goalScoring: 0.2,
  attack: 0.135,
  midfield: 0.133,
  defense: 0.133,
  fitness: 0.1,
}

export const calculatePlayerScore = player => {
  return (player.totalScore =
    player.gameKnowledgeScore * WEIGHTS.gameKnowledge +
    player.goalScoringScore * WEIGHTS.goalScoring +
    player.attackScore * WEIGHTS.attack +
    player.midfieldScore * WEIGHTS.midfield +
    player.defenseScore * WEIGHTS.defense +
    player.fitnessScore * WEIGHTS.fitness)
}

export const calculateTeamStats = team => {
  const stats = team.players.reduce(
    (acc, player) => {
      const playerScore = calculatePlayerScore(player)

      return {
        totalScore: acc.totalScore + playerScore,
        totalGameKnowledgeScore:
          acc.totalGameKnowledgeScore + player.gameKnowledgeScore,
        totalGoalScoringScore:
          acc.totalGoalScoringScore + player.goalScoringScore,
        totalAttackScore: acc.totalAttackScore + player.attackScore,
        totalMidfieldScore: acc.totalMidfieldScore + player.midfieldScore,
        totalDefenseScore: acc.totalDefenseScore + player.defenseScore,
        totalFitnessScore: acc.totalFitnessScore + player.fitnessScore,
        genderCount: {
          ...acc.genderCount,
          [player.gender]: (acc.genderCount[player.gender] || 0) + 1,
        },
      }
    },
    {
      totalScore: 0,
      totalGameKnowledgeScore: 0,
      totalGoalScoringScore: 0,
      totalAttackScore: 0,
      totalMidfieldScore: 0,
      totalDefenseScore: 0,
      totalFitnessScore: 0,
      genderCount: { male: 0, female: 0, nonBinary: 0 },
    }
  )
  return { ...team, ...stats }
}
