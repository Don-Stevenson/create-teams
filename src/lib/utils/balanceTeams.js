// Add weights at the top of the file
const WEIGHTS = {
  gameKnowledge: 0.2,
  goalScoring: 0.2,
  attack: 0.135,
  midfield: 0.133,
  defense: 0.133,
  fitness: 0.1,
}

// Add randomization to prevent identical team formations
const fudge = score => score + (Math.random() - 0.5) * 1.4

// Helper function to calculate weighted player score
const calculatePlayerScore = player => {
  return (
    player.gameKnowledgeScore * WEIGHTS.gameKnowledge +
    player.goalScoringScore * WEIGHTS.goalScoring +
    player.attackScore * WEIGHTS.attack +
    player.midfieldScore * WEIGHTS.midfield +
    player.defenseScore * WEIGHTS.defense +
    player.fitnessScore * WEIGHTS.fitness
  )
}

// Helper function to calculate fudged player score for sorting
const calculateFudgedPlayerScore = player => {
  return fudge(calculatePlayerScore(player))
}

function balanceTeams(players, numTeams) {
  // Validate input
  if (!players || !Array.isArray(players)) {
    console.error('Invalid players data:', players)
    throw new Error('Invalid players data')
  }
  if (!numTeams || numTeams < 2) {
    console.error('Invalid number of teams:', numTeams)
    throw new Error('Invalid number of teams')
  }

  // Ensure all required player properties exist and are of correct type
  const validPlayers = players.filter(player => {
    // Convert all numeric fields to numbers
    const gameKnowledgeScore = Number(player.gameKnowledgeScore)
    const goalScoringScore = Number(player.goalScoringScore)
    const attackScore = Number(player.attackScore)
    const midfieldScore = Number(player.midfieldScore)
    const defenseScore = Number(player.defenseScore)
    const fitnessScore = Number(player.fitnessScore)

    // Check if any numeric fields are NaN
    if (
      isNaN(gameKnowledgeScore) ||
      isNaN(goalScoringScore) ||
      isNaN(attackScore) ||
      isNaN(midfieldScore) ||
      isNaN(defenseScore) ||
      isNaN(fitnessScore)
    ) {
      console.warn('Player has invalid numeric scores:', {
        player,
        scores: {
          gameKnowledgeScore,
          goalScoringScore,
          attackScore,
          midfieldScore,
          defenseScore,
          fitnessScore,
        },
      })
      return false
    }

    // Check if gender is valid
    if (!['male', 'female', 'nonBinary'].includes(player.gender)) {
      console.warn('Player has invalid gender:', {
        player,
        gender: player.gender,
      })
      return false
    }

    // Check if isPlayingThisWeek is true (handle both boolean and string values)
    const isPlaying =
      player.isPlayingThisWeek === true ||
      player.isPlayingThisWeek === 'true' ||
      player.isPlayingThisWeek === 1 ||
      player.isPlayingThisWeek === '1'
    if (!isPlaying) {
      console.warn('Player is not marked as playing this week:', {
        player,
        isPlayingThisWeek: player.isPlayingThisWeek,
        type: typeof player.isPlayingThisWeek,
      })
      return false
    }

    // Create a clean player object with all required fields
    const cleanPlayer = {
      name: player.name,
      gameKnowledgeScore,
      goalScoringScore,
      attackScore,
      midfieldScore,
      defenseScore,
      fitnessScore,
      gender: player.gender,
      isPlayingThisWeek: true,
    }

    // Update the original player object with cleaned values
    Object.assign(player, cleanPlayer)

    return true
  })

  if (validPlayers.length === 0) {
    throw new Error('No valid players provided')
  }

  // Initialize teams
  const teams = Array.from({ length: numTeams }, () => ({
    players: [],
    totalScore: 0,
    totalGameKnowledgeScore: 0,
    totalGoalScoringScore: 0,
    totalAttackScore: 0,
    totalMidfieldScore: 0,
    totalDefenseScore: 0,
    fitnessScore: 0,
    genderCount: {
      male: 0,
      female: 0,
      nonBinary: 0,
    },
  }))

  // Separate players by gender to distribute women and non-binary first
  const femaleAndNonBinaryPlayers = validPlayers.filter(player =>
    ['female', 'nonBinary'].includes(player.gender)
  )
  const malePlayers = validPlayers.filter(player => player.gender === 'male')

  // Sort each group by fudged total score in descending order
  femaleAndNonBinaryPlayers.sort(
    (a, b) => calculateFudgedPlayerScore(b) - calculateFudgedPlayerScore(a)
  )
  malePlayers.sort(
    (a, b) => calculateFudgedPlayerScore(b) - calculateFudgedPlayerScore(a)
  )

  // Combine lists so female and non-binary players are added first
  const sortedPlayers = [...femaleAndNonBinaryPlayers, ...malePlayers]

  // Distribute players to teams with improved balancing
  for (let i = 0; i < sortedPlayers.length; i++) {
    const player = sortedPlayers[i]

    // Find teams with the minimum number of players
    const eligibleTeams = teams.filter(
      team =>
        team.players.length === Math.min(...teams.map(t => t.players.length))
    )

    // Among eligible teams, find the one with the lowest total score
    const targetTeam = eligibleTeams.reduce((min, team) =>
      team.totalScore < min.totalScore ? team : min
    )

    // Add player to the team (use actual score, not fudged score for team totals)
    targetTeam.players.push(player)
    targetTeam.totalScore += calculatePlayerScore(player)
    targetTeam.totalGameKnowledgeScore += player.gameKnowledgeScore
    targetTeam.totalGoalScoringScore += player.goalScoringScore
    targetTeam.totalAttackScore += player.attackScore
    targetTeam.totalMidfieldScore += player.midfieldScore
    targetTeam.totalDefenseScore += player.defenseScore
    targetTeam.fitnessScore += player.fitnessScore
    targetTeam.genderCount[player.gender]++
  }

  // Calculate final team stats
  const finalTeams = teams.map(team => ({
    ...team,
    totalScore: Number(team.totalScore.toFixed(2)),
    totalGameKnowledgeScore: Number(team.totalGameKnowledgeScore.toFixed(2)),
    totalGoalScoringScore: Number(team.totalGoalScoringScore.toFixed(2)),
    totalAttackScore: Number(team.totalAttackScore.toFixed(2)),
    totalMidfieldScore: Number(team.totalMidfieldScore.toFixed(2)),
    totalDefenseScore: Number(team.totalDefenseScore.toFixed(2)),
    fitnessScore: Number(team.fitnessScore.toFixed(2)),
    genderCount: {
      male: team.genderCount.male,
      female: team.genderCount.female,
      nonBinary: team.genderCount.nonBinary,
    },
  }))

  return { teams: finalTeams, totalPlayersPlaying: validPlayers.length }
}

export default balanceTeams
