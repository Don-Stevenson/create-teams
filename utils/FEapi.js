// utils/api.js
import axios from 'axios'

// Create axios instance
const api = axios.create({
  baseURL: process.env.NEXT_PUBLIC_API_URL || 'http://localhost:5050/api',
  withCredentials: true,
  headers: {
    'Content-Type': 'application/json',
    Accept: 'application/json',
  },
})

// Request interceptor
api.interceptors.request.use(
  config => {
    // Only stringify if data exists and is an object
    if (config.data && typeof config.data === 'object') {
      // Ensure arrays are properly stringified
      if (Array.isArray(config.data.players)) {
        config.data = {
          ...config.data,
          players: config.data.players,
        }
      }
      config.data = JSON.stringify(config.data)
    }
    return config
  },
  error => Promise.reject(error)
)

// Response interceptor
api.interceptors.response.use(
  response => response,
  async error => {
    const originalRequest = error.config

    // Handle 401 Unauthorized errors
    if (error.response?.status === 401 && !originalRequest._retry) {
      originalRequest._retry = true

      try {
        const authCheck = await api.get('/auth/check')

        if (authCheck.status === 200) {
          return api(originalRequest)
        }
      } catch (authError) {
        window.location.href = '/login'
        return Promise.reject(authError)
      }
    }

    console.error('API Error:', error.response)
    return Promise.reject(error)
  }
)

export const login = async (username, password) => {
  try {
    const response = await api.post('/login', { username, password })
    return response.data
  } catch (error) {
    console.error('Login Error:', error)
    throw error
  }
}

export const logout = async () => {
  try {
    const response = await api.post('/logout')
    return response.data
  } catch (error) {
    console.error('Logout Error:', error)
    throw error
  }
}

export const checkAuth = async () => {
  try {
    const response = await api.get('/auth/check')
    return response.data
  } catch (error) {
    throw error
  }
}

export default api
