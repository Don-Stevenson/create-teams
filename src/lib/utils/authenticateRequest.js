import { cookies } from 'next/headers'
import jwt from 'jsonwebtoken'
import { isBlacklisted } from './tokenBlacklist'
import { isSessionValid } from './sessionStore'

export async function authenticateRequest(request) {
  const cookieStore = cookies()
  let token = cookieStore.get('token')?.value

  // If no cookie token, check for Authorization header (Bearer token)
  if (!token) {
    const authHeader = request.headers.get('authorization')
    if (!authHeader) {
      throw new Error('Unauthenticated request')
    }
    token = authHeader.split(' ')[1] // Format: "Bearer TOKEN"
  }

  if (!token) {
    throw new Error('Unauthenticated request')
  }

  // Check if token is blacklisted
  const blacklisted = isBlacklisted(token)
  if (blacklisted) {
    throw new Error('Token has been invalidated')
  }

  // Check if session is valid
  const sessionValid = isSessionValid(token)
  if (!sessionValid) {
    throw new Error('Session has been invalidated')
  }

  try {
    const decoded = jwt.verify(token, process.env.JWT_SECRET)
    return { userId: decoded.userId, token }
  } catch (error) {
    if (error.name === 'TokenExpiredError') {
      throw new Error('Token has expired')
    }
    if (error.name === 'JsonWebTokenError') {
      throw new Error('Unauthenticated request')
    }
    throw error
  }
}

export function handleAuthError(error) {
  if (
    error.message.includes('Unauthenticated') ||
    error.message.includes('Token') ||
    error.message.includes('Session')
  ) {
    return { status: 401, message: error.message }
  }
  return { status: 500, message: 'Server error' }
}
