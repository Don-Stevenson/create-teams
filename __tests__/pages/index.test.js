import { render, screen, waitFor, act } from '@testing-library/react'
import '@testing-library/jest-dom'
import BalanceTeamsPage from '../../src/app/page'
import { checkAuth } from '../../utils/FEapi'

jest.mock('next/navigation', () => ({
  useRouter: () => ({
    push: jest.fn(),
  }),
}))

jest.mock('../../utils/FEapi', () => ({
  checkAuth: jest.fn(),
}))

describe('Home Page / balance teams', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders without crashing', async () => {
    checkAuth.mockResolvedValue(true)

    await act(async () => {
      render(<BalanceTeamsPage />)
    })

    expect(screen.getByText('Checking authentication...')).toBeInTheDocument()
  })

  it('shows checking authentication message initially', async () => {
    checkAuth.mockResolvedValue(true)

    await act(async () => {
      render(<BalanceTeamsPage />)
    })

    expect(screen.getByText('Checking authentication...')).toBeInTheDocument()
  })

  it('calls checkAuth on mount', async () => {
    checkAuth.mockResolvedValue(true)

    await act(async () => {
      render(<BalanceTeamsPage />)
    })

    // In development mode with React.StrictMode, effects run twice
    expect(checkAuth).toHaveBeenCalledWith()
  })

  it('handles authentication success', async () => {
    checkAuth.mockResolvedValue(true)

    await act(async () => {
      render(<BalanceTeamsPage />)
    })

    await waitFor(() => {
      expect(checkAuth).toHaveBeenCalled()
    })
  })

  it('handles authentication failure', async () => {
    checkAuth.mockResolvedValue(false)

    await act(async () => {
      render(<BalanceTeamsPage />)
    })

    await waitFor(() => {
      expect(checkAuth).toHaveBeenCalled()
    })
  })

  it('handles authentication error', async () => {
    const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {})
    checkAuth.mockRejectedValue(new Error('Network error'))

    await act(async () => {
      render(<BalanceTeamsPage />)
    })

    await waitFor(() => {
      expect(checkAuth).toHaveBeenCalled()
    })

    consoleSpy.mockRestore()
  })
})
