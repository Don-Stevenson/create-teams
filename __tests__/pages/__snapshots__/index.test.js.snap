// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page / balance teams matches snapshot 1`] = `
<div>
  <div
    data-testid="mock-layout"
  >
    <div
      class="flex-col mx-auto px-4 py-8"
    >
      <h1
        class="text-3xl font-bold mb-8 text-loonsRed print:hidden"
      >
        Create Teams
      </h1>
      <div
        data-testid="mock-create-teams"
      >
        CreateTeams Component
      </div>
    </div>
  </div>
</div>
`;
